pkg/progress/supermin.go

```diff
@@ -0,0 +1,338 @@
+
+# We cannot put the host osbuild "store" on our /store dir even with
+# virtiofsd as there will be issues with uid remapping and selinux
+# labels.
+mkdir -p /host/store
+mount -t virtiofs osbuild_store /host/store
+# rsync/cp the caches so that our build is faster
+echo "Populate /store from host"
+mkdir /store
+rsync -a --exclude=./tmp /host/store/ /store
+
+# fetch/cache sources first so that the host cache gets updated
+echo "Fetching sources"
+osbuild \
+  --cache /store \
+  /output/manifest.json
+# rsync
+rsync -a --exclude=./tmp /store/ /host/store
+
+echo "Running osbuild"
+osbuild \
+  --export %s \
+  --output-directory /output \
+  --cache /store \
+  /output/manifest.json
+
+echo "" > /output/%s
+
+# trigger clean shutdown via sysreq
+echo _suo > /proc/sysrq-trigger
+# shutdown is async so we need to sleep here or PID=0 dies
+sleep 999
+`
+
+func addInitTar(superminDir, superminInitScript string) error {
+	initTarF, err := os.Create(filepath.Join(superminDir, "init.tgz"))
+	if err != nil {
+		return err
+	}
+	defer initTarF.Close()
+	initTar := tar.NewWriter(initTarF)
+	defer initTar.Close()
+	if err := initTar.WriteHeader(&tar.Header{
+		Name: "init",
+		Size: int64(len(superminInitScript)),
+		Mode: 0755,
+	}); err != nil {
+		return err
+	}
+	if _, err := initTar.Write([]byte(superminInitScript)); err != nil {
+		return err
+	}
+	return nil
+}
+
+func superminPrepare(prepareDir, export, doneStampName string) error {
+	superminInitScript := fmt.Sprintf(superminInitScriptFmt, export, doneStampName)
+
+	// prepare supermin
+	err := util.RunCmdSync(
+		"supermin", "--prepare", "--use-installed",
+		// fundamental
+		"util-linux",
+		// convenient
+		"rsync",
+		// basic networking
+		"ca-certificates", "dhcp-client", "iproute",
+		// loop-device support
+		"kernel-modules",
+		// osbuild and friends
+		"osbuild", "osbuild-depsolve-dnf", "osbuild-lvm2", "osbuild-luks2", "osbuild-ostree",
+		// lvm
+		"lvm2",
+		// target dir
+		"-o", prepareDir,
+	)
+	if err != nil {
+		return fmt.Errorf("supermin prepare failed: %w", err)
+	}
+	if err := addInitTar(prepareDir, superminInitScript); err != nil {
+		return err
+	}
+	return nil
+}
+
+func superminBuild(prepareDir, buildDir string) error {
+	err := util.RunCmdSync(
+		"supermin",
+		"--build", prepareDir,
+		// XXX: what is the right size?
+		"--size", "40G",
+		"-f", "ext2",
+		"-o", buildDir,
+	)
+	if err != nil {
+		return fmt.Errorf("supermin-build failed: %w", err)
+	}
+	return nil
+}
+
+func waitForFiles(maxWait time.Duration, paths ...string) error {
+	start := time.Now()
+	for {
+		var missingPaths []string
+		for _, path := range paths {
+			if _, err := os.Stat(path); err != nil {
+				missingPaths = append(missingPaths, path)
+			}
+		}
+		if len(missingPaths) == 0 {
+			return nil
+		}
+		if time.Since(start) >= maxWait {
+			return fmt.Errorf("files missing after %v: %v", maxWait, missingPaths)
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
```

Comment:
I might prefer to do this with `time.Ticker` and a `context.WithDeadline` instead but since we don't *really* use either anywhere else I won't block on it.

Example:

```go
func waitForFiles(ctx context.Context, paths ...string) error {
	ticker := time.NewTicker(100 * time.Millisecond)
    defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			// work
		case <-ctx.Done():
			// cancelled due to timeout
		default:
            // wait
		}
	}
}
```

Reply:
Yea, also it's not *really* worth the effort since we're not using context(s) anywhere else in our codebase(s). If we were it'd be nice that the contexts inherit everywhere and we can cancel them from all the way up the stack.

Let's skip this then.